### ISSUE TITLE
How to set a discovery to active

### ISSUE THREAD

**Issue by NathanDotTo** [CiscoDevNet/uniq#29]
Hi

I am trying to figure out how to use the DiscoveryAPI. The code I have put together is below, but it is not quite working.

I am getting " Bad Request for url: https://192.168.1.13/api/v1/discovery" when calling the client.discovery.updateDiscovery() function.

What should I be doing here please?

Many thanks

Nathan

```
import requests.exceptions

from uniq.apis.nb.client_manager import NbClientManager
from uniq.apis.nb.clients.inventory_manager_client.models.DiscoveryNIO import DiscoveryNIO


try:
  client = NbClientManager(
                server="192.168.1.13",
                username="admin",
                password="xxxxxx!",
                connect=True)

except requests.exceptions.HTTPError as exc_info:
  print(exc_info)
  exit(1)

discoveryNIO = DiscoveryNIO()
discoveryNIO.discoveryStatus = "active"

try:
  client.discovery.updateDiscovery(discovery = discoveryNIO, scope = "All")
except requests.exceptions.HTTPError as exc_info:
  print(exc_info)
  exit(1)

exit(0)
```

**Comment by himakoppu** [comment]
Hi Nathan,

After you create a discovery job, to make it active or inactive, you need the 'id' of that discovery job. 

First get all discoveries, and find the id of your discovery job. Like below 

```
discoveries = client.discovery.getDiscoveryByRange(startIndex=1, recordsToReturn=100, scope="All")
for discovery in discoveries.response:
    if discovery.name == discovery_name:
        discovery_id = discovery.id
        break
```

Next, once you get the discovery id,

```
update_request = {"discoveryStatus": "active", "id": discovery_id}
try:
    client.discovery.updateDiscovery(discovery=update_request, scope="All")
except requests.exceptions.HTTPError as exc_info:
    print(exc_info)
```

**Comment by NathanDotTo** [comment]
Thanks. So I forgot the ID. 

It also looks like the DiscoveryNIO is not actually used, as you are passing JSON, rather than an instance of a DiscoveryNIO.

Is that right?

**Comment by himakoppu** [comment]
Hi Nathan,

Most of APIC-EM's POST and PUT requests need dict/json in the request body. DiscoveryNIO is a request model for updateDiscovery(). It represents what key/value pairs comprise the dict/json in the request.

**Comment by NathanDotTo** [comment]
Thanks for the response. 

Do you expect users of the API to pass JSON, or DiscoveryNIO, instances or does it not matter?

Regards

Nathan

**Comment by ameyspatil** [comment]
@NathanDotTo all requests that are made should be made using json/python-dict.
The models like DiscoveryNIO will be used to create objects of, in the API response.

**Comment by NathanDotTo** [comment]
OK. The DiscoveryAPI says:

```
def updateDiscovery(self, **kwargs):
    """Updates an existing discovery specified by id - only for starting/stopping the discovery
    Args:
        discovery, DiscoveryNIO: discovery (required)
```

Which is why I was using a DiscoveryNIO instance. 

You might want to consider rephrasing the documentation.

Thanks

Nathan

**Comment by NathanDotTo** [comment]
For posterty's sake, here is the code that worked for me:

```
import requests.exceptions
from uniq.apis.nb.client_manager import NbClientManager
from uniq.apis.nb.clients.inventory_manager_client.models.DiscoveryNIO import DiscoveryNIO

try:
  client = NbClientManager(
                server="192.168.1.13",
                username="admin",
                password="xxxxx",
                connect=True)

  discovery_name = "a name"

  discoveries = client.discovery.getDiscoveryByRange(startIndex=1, recordsToReturn=100, scope="All")
  for discovery in discoveries.response:
    if discovery.name == discovery_name:
      discovery_id = discovery.id
      print("Discovery ID - " + discovery_id)
      break

  update_request = {"discoveryStatus": "active", "id": discovery_id}
  task_id = client.discovery.updateDiscovery(discovery=update_request, scope="All")
  print("Task ID - " + str(task_id))

  client.task_util.wait_for_task_success(task_id, timeout=3, poll_frequency=0.5)
  print("Task completed")
except requests.exceptions.HTTPError as exc_info:
  print(exc_info)
  exit(1)

exit(0)
```

**Comment by ameyspatil** [comment]
I apologize, I misspoke. Let me rephrase:
@NathanDotTo all requests that are made can be made using dicts apart from using model objects.
In the API response, however, there will be objects of models for nearly all the API endpoints.

So from the above snippet, this change would still work:

```

import requests.exceptions
from uniq.apis.nb.client_manager import NbClientManager
from uniq.apis.nb.clients.inventory_manager_client.models.DiscoveryNIO import DiscoveryNIO

try:
  client = NbClientManager(
                server="192.168.1.13",
                username="admin",
                password="xxxxx",
                connect=True)

  discovery_name = "a name"

  discoveries = client.discovery.getDiscoveryByRange(startIndex=1, recordsToReturn=100, scope="All")
  for discovery in discoveries.response:
    if discovery.name == discovery_name:
      discovery_id = discovery.id
      print("Discovery ID - " + discovery_id)
      break

  discoveryNIO = DiscoveryNIO()
  discoveryNIO.discoveryStatus = "active"
  discoveryNIO.id = discovery_id

  task_id = client.discovery.updateDiscovery(discovery=discoveryNIO, scope="All")
  print("Task ID - " + str(task_id))

  client.task_util.wait_for_task_success(task_id, timeout=3, poll_frequency=0.5)
  print("Task completed")
except requests.exceptions.HTTPError as exc_info:
  print(exc_info)
  exit(1)

exit(0)
```

**Comment by NathanDotTo** [comment]
Got it, thanks. That form worked for me too, so now my code matches yours.

Please feel free to add this is a sample.

Regards

Nathan

**Comment by ameyspatil** [comment]
@NathanDotTo we used your example, and added a new sample script file
https://github.com/CiscoDevNet/uniq/pull/35
